operator: Add recovery steps for ingester pods in zone-failure outages (#10057)
grafana/loki: operator/docs/user-guides/recover_ingester_pods_in_zone_failure_outages.md

---
title: "Playbook for recovering ingester pods in zone-failure outages"
description: "Playbook for recovering ingester pods in zone-failure outages"
lead: ""
date: 2023-07-25T08:48:45+00:00
lastmod: 2023-07-25T08:48:45+00:00
draft: false
images: []
menu:
  docs:
    parent: "user-guides"
weight: 100
toc: true
---

This document describes how to recover the ingester pods in the event of a zone failure.

> _Disclaimer:_ This document describes a recovery procedure by manually recreating the failed pods in another zone. Right now, we are doing this by deleting PVCs of the impacted pods from the failed zone, so it can be recreated in a different zone. This will cause data loss of the data in the PVC. To avoid actual data loss we always set the replication factor in loki to be 1 or higher so data is always replicated.

Reviewer: nit
```suggestion
_Disclaimer:_ This document describes a recovery procedure by manually recreating the failed pods in another zone. Right now, we are doing this by deleting PVCs of the impacted pods from the failed zone, so they can be recreated in a different zone. This will cause data loss of the data in the PVC. To avoid actual data loss we always set the replication factor in loki to be 1 or higher so data is always replicated.
```

Author: Done

> _Disclaimer:_ This document describes a recovery procedure by manually recreating the failed pods in another zone. Right now, we are doing this by deleting PVCs of the impacted pods from the failed zone, so they can be recreated in a different zone. This will cause data loss of the data in the PVC. To avoid actual data loss we always set the replication factor in loki to be 1 or higher so data is always replicated.

Reviewer 2: PVC - As a best practice, spell out acronyms on first use. AcronymFinder has 49 definitions for PVC.
Also, should `loki` be captialized here?

Author: Done

Reviewer 2: I'm not seeing where PVC is spelled out?  Assuming since we're talking about Kubernetes that we're talking about a PersistentVolumeClaim?  Could we explicitly state that somewhere please?

Author: Sorry missed the acronym expansion. I have added it now

## Ingester recovery during a Zone Failure

In a Kubernetes/Openshift cluster, a "zone failure" refers to a situation where nodes or resources in a specific availability zone become unavailable. An availability zone is a distinct location within a cloud provider's data center or region, designed to be isolated from failures in other zones to provide better redundancy and fault tolerance. When a zone failure occurs, it can lead to a loss of services or data if the cluster is not configured properly to handle such scenarios.

> This document outlines steps that can be taken to recover ingester pods when there is a zone failure. Ingester pods are deployed as a part of a Statefulset. The Ingester Statefulset also has PVCs associated with the pods. Each Ingester Pod and its associated PVCs are deployed in the same zone as the pod. This is done by using the storageclass.

Reviewer: I wasn't understanding the last two phrases, feel free to fix it and don't accept this suggestion
```suggestion
This document outlines steps that can be taken to recover ingester pods when there is a zone failure. Ingester pods are deployed as a part of a Statefulset. The Ingester Statefulset also has PVCs associated with the pods. Each Ingester Pod and its associated PVCs are deployed in the same zone.
```

Author: I had missed adding the Dynamic provisioning part. Added now. Does it make it more clear now?

> This document outlines steps that can be taken to recover ingester pods when there is a zone failure. Ingester pods are deployed as a part of a Statefulset. The Ingester Statefulset also has PVCs associated with the pods which are dynamically provisioned through the use of StorageClass. Each Ingester Pod and its associated PVCs are deployed in the same zone.

Reviewer 2: ```suggestion
This document outlines steps that can be taken to recover ingester pods when there is a zone failure. Ingester pods are deployed as a part of a StatefulSet. The Ingester StatefulSet also has PVCs associated with the pods which are dynamically provisioned through the use of StorageClass. Each Ingester Pod and its associated PVCs are deployed in the same zone.
```

Author: Done

> When a zone failure occurs in a cluster, the StatefulSet controller will automatically attempt to recover the affected pods in the failed zone. This document outlines the additional manual intervention required to make sure that the ingester pods are succesfully recreated in a new zone.

Reviewer 2: ```suggestion
When a zone failure occurs in a cluster, the StatefulSet controller will automatically attempt to recover the affected pods in the failed zone. This document outlines the additional manual intervention required to make sure that the Ingester pods are successfully recreated in a new zone.
```

Author: Done

> 

Reviewer 2: This bulleted list needs some kind of an introduction.  Is this the list of manual configurations that users need to make?  If yes, then I'd change line 26 to read "The following is a list of additional manual interventions required to make sure that the Ingester pods are successfully recreated in a new zone."

If this is a list of ways that the StatefulSet Controller attempts to recover, then I'd delete ` This document outlines the additional manual intervention required to make sure that the Ingester pods are successfully recreated in a new zone.` as it doesn't fit here.

Author: Formatted the recovery steps that occurs when a zone failure happens. And added a sub section for the specific list of manual steps required to recover the Loki stateful pods - https://github.com/grafana/loki/pull/10057/files#diff-af523907b539b8c2dc834a5d314e3c03f5f2efdd1d583ca99a1b9082ac96dd1dR57

* Detect Zone Failure - The control plane and cloud provider integration should mark nodes in the failed zone.

* Reschedule Pods: The Ingester StatefulSet controller will automatically attempt to reschedule the pods that were running in the failed zone to nodes in another zone.
  

> * Since the Statefulset has PVCs which are also in the failed zone, automatic reschedule of the Ingester pods will not work. Manual intervention is required at this point.

Reviewer: Should we give more detail on why the StatefulSet having PVCs will make the rescheduling not work? I feel like it makes sense just so users are aware that it's a k8s limitation and not a limitation of out stack

Author: Added more comments and a link to a document on official Kubernetes documentation. Does this help?

> * Since the Statefulset has PVCs which are also in the failed zone, automatic reschedule of the Ingester pods to a different zone will not work. See - [Storage access for zones](https://kubernetes.io/docs/setup/best-practices/multiple-zones/#storage-access-for-zones). Manual intervention is required at this point to delete the old pvcs in the failed zone to allow succesful recreation of the Ingester Pod & Pvc in the new zone.

Reviewer 2: Question, why are you directing users to the Storage access for zones page?  
If it's just general knowledge, then `for more information about storage access for zones, see the [Kubernetes documentation].`
If it is for configuration, then `To learn how to configure storage access for zones, see the [Kubernetes documentation].`

> * Since the Statefulset has PVCs which are also in the failed zone, automatic reschedule of the Ingester pods to a different zone will not work. See - [Storage access for zones](https://kubernetes.io/docs/setup/best-practices/multiple-zones/#storage-access-for-zones). Manual intervention is required at this point to delete the old pvcs in the failed zone to allow succesful recreation of the Ingester Pod & Pvc in the new zone.

Reviewer 3: ```suggestion
* Since the Statefulset has PVCs which are also in the failed zone, automatic reschedule of the Ingester pods to a different zone will not work. See - [Storage access for zones](https://kubernetes.io/docs/setup/best-practices/multiple-zones/#storage-access-for-zones). Manual intervention is required at this point to delete the old PVCs in the failed zone to allow succesful recreation of the Ingester Pod & PVC in the new zone.
```

Author: Done

> * Since the Statefulset has PVCs which are also in the failed zone, automatic reschedule of the Ingester pods to a different zone will not work. See - [Storage access for zones](https://kubernetes.io/docs/setup/best-practices/multiple-zones/#storage-access-for-zones). Manual intervention is required at this point to delete the old pvcs in the failed zone to allow succesful recreation of the Ingester Pod & Pvc in the new zone.

Author: @JStickler : Updated it, as per your suggestion

>     Example showing an ingester pod that is in a pending state, after the statefulset has unsuccesfully tried to reschedule it to a different zone.

Reviewer 3: ```suggestion
    Example showing an Ingester pod that is in a pending state, after the StatefulSet has unsuccessfully tried to reschedule it to a different zone.
```

Author: Done

    Ingester-1 is in Pending state because it cannot be scheduled in a new node in a different zone:

>     ```console

Reviewer 2: Format the command and the result as separate code blocks.  That way customers can copy/paste just the command.

Author: Done

    oc get pods -n openshift-logging
    NAME                                           READY   STATUS    RESTARTS   AGE
    lokistack-dev-ingester-0                       1/1     Running   0          4m21s
    lokistack-dev-ingester-1                       0/1     Pending   0          44s
    ```

    This is because the pvc for ingester-1 is in the old zone

>     ```console

Reviewer 2: Same, format the command and the result as separate code blocks.  That way customers can copy/paste just the command.

Author: Done

    oc get pvc -n openshift-logging
    NAME                                    STATUS    VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
    storage-lokistack-dev-ingester-0        Bound     pvc-1be8a2c2-f594-4ce7-9139-d57de6e06ce9   10Gi       RWO            gp3-csi        12m
    storage-lokistack-dev-ingester-1        Pending                                                                        gp3-csi        8m24s
    wal-lokistack-dev-ingester-0            Bound     pvc-0239dca6-cf2c-4b67-94a8-e2e0fb94509d   150Gi      RWO            gp3-csi        12m
    wal-lokistack-dev-ingester-1            Pending                                                                        gp3-csi        8m24s
    ```

>     Delete the PVCs, followed by the pod. The pod and new pvcs should now be recreated in an available zone because the statefulset has a set number of replicas

Reviewer 3: ```suggestion
    Delete the PVCs, followed by the pod. The pod and new PVCs should now be recreated in an available zone because the StatefulSet has a set number of replicas
```

    ```console
    oc delete pvc storage-lokistack-dev-ingester-1 -n openshift-logging
    oc delete pvc wal-lokistack-dev-ingester-1 -n openshift-logging
    oc delete pod lokistack-dev-ingester-1 -n openshift-logging
    ```

> These steps should be followed for all ingester pods that are in the failed zone.

Reviewer 3: ```suggestion
These steps should be followed for all Ingester pods that are in the failed zone.
```

Author: Done

> As discussed in the disclaimer above. This process will delete the PVC in the failed zone and the data held there. To avoid complete data loss the replication factor in the lokistack CR should always be set to a value greater than 0. This ensures that loki is replicating the data and even if a zone is lost there should be already be copies of the data in another zone.

Reviewer 2: ```suggestion
This process will delete the PVC in the failed zone and the data held there. To avoid complete data loss the replication factor in the LokiStack custom resource (CR) should always be set to a value greater than 0. This ensures that Loki is replicating the data and even if a zone is lost there should be already be copies of the data in another zone.
```
